Add Azure REST API client to manage Windows virtual machines.

jankeromnes opened the pull request:
Fixes #268.

I'm also sneaking in the following drive-by changes:

- Refactor stream data processing to use Buffer.concat() (following @bnjbvr's nit suggestion)
- Support 'aux' channel in Docker Remote API response streams, fix #251 (because I came across the code during the refactor)
- Remove annoying package-lock.json file from repository (it always causes conflicts and I don't find it useful)
- Add annoying package-lock.json file to .gitignore (see above)
- Remove deprecated in-app proxy handlers now that app and proxies were separated (stolen from #188 because review is taking too long over there 😋 also it mitigates a WebSocket DDoS found by @Snyk)

jankeromnes commented:
@nt1m or @bnjbvr could you please have a look at this, whoever has time for it first? 😄 

This is an admin-only feature (so no need to spend a lot of effort on UX tweaks), and here is what it looks like (works locally in my container):

![azure-admin](https://user-images.githubusercontent.com/599268/35977320-1553fac2-0ce3-11e8-8ea5-56a3301809db.png)

bnjbvr commented on api/admin-api.js:
```diff
@@ -13,6 +14,80 @@ const adminAPI = module.exports = selfapi({
+
+    let json = '';
+    request.on('data', chunk => {
+      json += String(chunk);
```
nit: potential hard to debug encoding issue here; use `Buffer` for the initial value, `Buffer.concat` here, and then JSON.parse the `buffer.toString()` at the end

bnjbvr commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  // You can customize these values in `./db.json` or via `/admin/`.
+  const { credentials } = azure;
```
Considering that the defaulting above has no effect (not saved), you could just do `const credentials = azure.credentials || {};`

jankeromnes replied:
Actually no, because we need `azure` to be a real DB reference (`db.save()` only serves to dump the DB to disk, but any changes to something from `db.get(key)` will work anywhere, since it's a direct in-memory DB reference).

With your suggestion, templates retrieving the `azure` object from the DB (e.g. the integrations admin page) won't have empty credential values (the default value for a non-existant `db.get(key)` is `{}`).

bnjbvr commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+  const { credentials } = azure;
+
+  // "Application ID" or "Client ID".
+  if (!credentials.clientId) {
```
if it's a numerical value, note it could be falsy for 0.

jankeromnes replied:
I suppose that `0` is not a valid Azure Application ID anyway (they have a certain minimum length), so that's ok.

bnjbvr commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  // "Domain" or "Directory ID" or "Tenant ID".
+  if (!credentials.tenantId) {
```
ditto

jankeromnes replied:
ditto

bnjbvr commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  // "Azure Subscription ID".
+  if (!credentials.subscriptionId) {
```
ditto

jankeromnes replied:
ditto

bnjbvr commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+  // "Azure Subscription ID".
+  if (!credentials.subscriptionId) {
+    credentials.subscriptionId = '';
+  }
```
you need to `db.save()` here, right?

jankeromnes replied:
Actually no. Here we're in the case were, immediately when started, the server sets a few empty default values. We don't need to `db.save()` them, because even if the server crashes, upon reboot the exact same values will be set again.

(Also, calling `db.save()` at approximately the same time as a server crash exposes us to wiping the `db.json` file. So calling `db.save()` on startup is both pointless *and* dangerous, e.g. if you have obvious accidental `TypeErrors` somewhere.)

nt1m commented on lib/azure.js:
```diff
@@ -0,0 +1,79 @@
+    );
+  });
+
+  const computeClient = new ComputeManagementClient(credentials, subscriptionId);
```
You can do:
```
return new ComputeManagementClient(credentials, subscriptionId);
```

jankeromnes commented:
@bnjbvr Thanks a lot for your review! I've addressed your nits (everywhere applicable, e.g. by refactoring all our stream processing snippets to use `Buffer.concat()`) and replied to the ones I decided not to act upon.

I also fixed a drive-by bug, #251. Could you please have a look at my new changes? Thanks! 😀

nt1m commented on api/admin-api.js:
```diff
@@ -13,6 +14,79 @@ const adminAPI = module.exports = selfapi({
+    }
+  },
+
+  examples: [],
```
Does this need examples for tests ?

jankeromnes replied:
Eventually yes, but I don't want to spend time on that yet. I think it's ok to omit examples for now.

nt1m commented on static/js/admin.js:
```diff
@@ -70,6 +70,23 @@ Array.map(document.querySelectorAll('button[data-action]'), function (button) {
+  button.addEventListener('click', function (event) {
+    var url = '/api/admin/azure/virtualmachines';
+    pre.textContent = 'Fetching virtual machines…';
+    window.fetchAPI('GET', url, null, function (error, virtualMachines) {
```
nit: janitor/janitor-new.js just use `fetchAPI`, while admin.js uses `window.fetchAPI`. Please be consistent. (I prefer just `fetchAPI` because it's shorter).

jankeromnes replied:
Actually, this file has a lot of occurrences of `window.fetchAPI`. It was to silence Editor gutter warnings in Cloud9 IDE.

If you don't mind, I'd like to leave this small inconsistency between files for now, but we should definitely restore consistency as a follow-up.

nt1m replied:
Sounds good !

jankeromnes commented:
Notes:

- I told @Snyk to ignore all `socket.io` vulnerabilities forever, because we don't use `camp`'s optional `socket.io` module. (Since we never read the `camp.io` [magic property](https://github.com/espadrine/sc/blob/1ebcc6ca201427593826d44dcf6cdbf909f702f5/lib/camp.js#L417-L431), we should never even load `socket.io` modules into memory. So in theory, it's just dead weight in our NPM dependencies.)
- I also told @Snyk to ignore the `ws` DDoS vulnerability for 30 days. This one does affect us, although just our Docker host proxies (i.e. `join.js`, and not our main web server `app.js`, since I've just deleted the deprecated WebSocket proxy there). This should resolve itself when @espadrine gets a change to upgrade `camp` to `ws 4.0.0` and/or to enable https://greenkeeper.io (see https://github.com/espadrine/sc/issues/66).

jankeromnes commented:
Thanks a lot for the reviews! 💯
